### Importing libraries

In [124]:
import pandas as pd
import numpy as np

## Question 1
### What's the version of NumPy that you installed

In [125]:
np.__version__

'1.22.3'

### Answer 1
**1.22.3**

### Importing the car price dataset

In [126]:
car_price = pd.read_csv('../datasets/Car-Price-Dataset.csv')

## Question 2
### How many records are in the dataset? (Car Price dataset)

In [127]:
car_price.shape

(11914, 16)

### Answer 2:
**11914**

## Question 3
### Who are the most frequent car manufacturers (top-3) according to the dataset?

In [128]:
car_price['Make'].value_counts()

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
Alfa Romeo          5
McLaren             5
Spyker    

### Answer 3
**Chevrolet, Ford, Volkswagen**

## Question 4
### What's the number of unique Audi car models in the dataset?

In [129]:
car_price[car_price['Make']== 'Audi'].Model.nunique()

34

### Answer 4
**34**

## Question 5
### How many columns in the dataset have missing values?

In [130]:
cols = car_price.columns[car_price.isna().any()]
print(len(cols))

5


### Answer 5
**5**

## Question 6
### 1. Find the median value of "Engine Cylinders" column in the dataset.

### 2. Next, calculate the most frequent value of the same "Engine Cylinders".

### 3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.

### 4. Now, calculate the median value of "Engine Cylinders" once again. Has it changed?

In [131]:
# median of Engine Cylinders column
med = car_price['Engine Cylinders'].median()
print(med)

6.0


In [132]:
# Mode of Engine Cylinders column
mode = car_price['Engine Cylinders'].mode()
print(mode)

0    4.0
Name: Engine Cylinders, dtype: float64


In [133]:
# Filling missing values in Engine Cylinders column
car_price['Engine Cylinders'] = car_price['Engine Cylinders'].fillna(4.0)

In [134]:
# median of Engine Cylinders column
med = car_price['Engine Cylinders'].median()
print(med)

6.0


### Answer 6
**No**

## Question 7
### 1. Select all the "Lotus" cars from the dataset.
### 2. Select only columns "Engine HP", "Engine Cylinders".
### 3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
### 4. Get the underlying NumPy array. Let's call it X.
### 5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
### 6. Invert XTX.
### 7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
### 8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w. What's the value of the first element of w?

In [135]:
# Selecting all the Lotus cars from dataset
lotus = car_price[car_price['Make']=='Lotus']

In [136]:
# Selectin only Engine Hp and Engine Cylinders column
lotus = lotus[['Engine HP','Engine Cylinders']]

In [137]:
# Dropping all the duplicates
lotus_dup = lotus.drop_duplicates()

In [138]:
# The dataframe left with only 9 rows
lotus_dup

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [139]:
# Getting the underlying numpy array
X = np.array(lotus_dup)

In [140]:
# Matrix multiplication of X trasnpose and X
XTX = np.matmul(X.T,X)

In [141]:
# Calculating inverse of XTX
inv_XTX = np.linalg.inv(XTX)

In [142]:
# Creating array y
y = [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]

In [143]:
# Multiply the inverse of XTX with the transpose of X
mul = np.dot(inv_XTX,X.T)

In [144]:
# multiplying the result with y
w = np.dot(mul,y)

In [145]:
w[0]

4.594944810094551

### Answer 7
**4.5949**